### PR TITLE
net: lwm2m: Protect send() calls with a mutex 

### PR DESCRIPTION
**Pull request by rlubos**
Add mutex protection for `send()` calls on the same socket. Although LwM2M engine uses cooperative threads, the internal `send()` implementation might trigger context switch when it calls a kernel function, therefore resulting in `send()` call being entered from both the LwM2M thread and the retransmit work.

Additionally, fix a small issue spotted when running FOTA on the `lwm2m_client` sample with both, IPv4 and IPv6 enabled. In case one socket communicated over IPv4 and the other one over IPv6, the communication fails on IPv6  socket due to an insufficient `sockaddr` structure size provided.

### CODE REVIEW

**Review by jukkar — APPROVED** [review]

File: subsys/net/lib/lwm2m/lwm2m_engine.c [hunk 1]
```diff
@@ -3769,11 +3776,22 @@ static void retransmit_request(struct k_work *work)
 
 	LOG_INF("Resending message: %p", msg);
 	msg->send_attempts++;
+
+	sys_mutex_lock(&client_ctx->send_lock, K_FOREVER);
+
+	if (msg->ctx == NULL) {
+		LOG_INF("Response for %p already handled", msg);
+		sys_mutex_unlock(&client_ctx->send_lock);
```
Comment (pabigot) [inline subsys/net/lib/lwm2m/lwm2m_engine.c:3784]: Similarly if you make this `goto next_locked` and....

File: subsys/net/lib/lwm2m/lwm2m_engine.c [hunk 2]
```diff
@@ -3769,11 +3776,22 @@ static void retransmit_request(struct k_work *work)
...
+		goto next;
+	}
+
 	if (send(msg->ctx->sock_fd, msg->cpkt.data, msg->cpkt.offset, 0) < 0) {
 		LOG_ERR("Error sending lwm2m message: %d", -errno);
 		/* don't error here, retry until timeout */
 	}
 
+	sys_mutex_unlock(&client_ctx->send_lock);
```
Comment (pabigot) [inline subsys/net/lib/lwm2m/lwm2m_engine.c:3793]: ... make this
```
next_locked:
      sys_mutex_unlock(&client_ctx->send_lock);
```
the flow will be cleaner.  In my opinion, at least.

**Review by pabigot — CHANGES_REQUESTED** [review]
Seems OK but I'd rather see a cleaner pairing of lock with unlock.

**Comment by rlubos** [comment]
@pabigot Thank you for the feedback, I've simplified the mutex handling as suggested.

**Review by pabigot — APPROVED** [review]

**Pull request merged**